twake/backend/node/src/services/user/web/controller.ts

```diff
@@ -6,40 +6,200 @@ import { ResourceListResponse } from "../../../services/types";
 import { ResourceDeleteResponse } from "../../../services/types";
 import { ResourceCreateResponse } from "../../../services/types";
 import { ResourceGetResponse } from "../../../services/types";
-import { UsersServiceAPI } from "../api";
+import { CompaniesServiceAPI, UsersServiceAPI } from "../api";
 
 import User from "../entities/user";
-import { UserListQueryParameters, UserParameters } from "./types";
+import {
+  UserCompanyObject,
+  UserListQueryParameters,
+  UserParameters,
+  UserObject,
+  UserCompanyRole,
+  UserCompanyStatus,
+  CompanyShort,
+  CompanyObject,
+  CompanyStatsObject,
+} from "./types";
+import assert from "assert";
+import Company from "../entities/company";
+import CompanyUser from "../entities/company_user";
 
-export class UsersCrudController implements CrudController<ResourceGetResponse<User>, ResourceCreateResponse<User>, ResourceListResponse<User>, ResourceDeleteResponse> {
-  constructor(protected service: UsersServiceAPI) {}
+export class UsersCrudController
+  implements
+    CrudController<
+      ResourceGetResponse<UserObject>,
+      ResourceCreateResponse<User>,
+      ResourceListResponse<UserObject>,
+      ResourceDeleteResponse
+    > {
+  constructor(protected service: UsersServiceAPI, protected companyService: CompaniesServiceAPI) {
+    assert(service, "service is not inited");
+    assert(companyService, "companyService is not inited");
+    this.companiesCache = {};
+  }
+
+  companiesCache = {} as { [key: string]: Company };
```

Comment:
It's cache for an instance just to reduce requests for the same company for two users. It invalidates on destruction.

obj → map: ok

Reply:
@romanesko how do you deal with this when we have multiple NodeJS instances and a load balancer? What happens when a company is deleted (true, this is not so frequent), or updated, etc? This may not cause problems here but can cause some in some situation.
This is why cache in distributed environment is complex, and in our case, should be managed at the ORM level.

Reply:
Cache in single node is still a good idea. But yes, it should be moved to the ORM itself good idea.

Reply:
It's not a cache in common sense.
It leaves within only a one request, it is just a trick, to economize a bit database queries

look, I have async requests:
```js
const [currentUserCompanies, requestedUserCompanies] = await Promise.all(
      [context.user.id, request.params.id].map(id =>
        this.service.getUserCompanies({ id }).then(a => a.getEntities()),
      ),
 );
    
```    
the companies may be common for both users, in this case if i get a company for one user I don't have to do query db for another (cause it is should be the same).

Reply:
Did I misunderstand something?

Reply:
Soo I think I got it now —  don't we create a new instance of controller for every request?